### PR TITLE
test.py topology Scylla REST API client

### PR DESCRIPTION
**Pull request by alecco**
- Separate `aiohttp` client code
- Helper to access Scylla server REST API
- Use helper both in `ScyllaClusterManager` (test.py process) and `ManagerClient` (pytest process)

As `ScyllaCluster` manages through IPs some APIs require to specify both.

### CODE REVIEW

File: test/pylib/rest_client.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                logger.info("wait_for_host_known [{%s}] timeout", expect_server_id)
+                return False
+            logger.debug("wait_for_host_known [{%s}] not found, sleeping {sleep}", expect_server_id)
+            await asyncio.sleep(sleep)
+
+    async def remove_node(self, initiator_ip: str, server_uuid: str) -> bool:
+        """Initiate remove node of server_uuid in initiator initiator_ip"""
+        resp = await self.cli.post("/storage_service/remove_node", params={"host_id": server_uuid},
+                                   host=initiator_ip)
+        logger.debug("remove_node status %s for %s", resp.status, server_uuid)
```
Comment (gusev-p) [inline test/pylib/rest_client.py:135]: `TCPRESTClient.post` already checks for 200 status code, so `resp.status == 200` is redundant. Seems that we can leave void as return type for `remove_node` and `wait_for_remove`, I'm ok with getting exception if something went wrong.

Regarding http server handling, maybe we can have general logic, which will return 500/log everything in case of exceptions?

File: test/pylib/scylla_cluster.py [hunk 2]
```diff
@@ -844,9 +848,24 @@ async def _cluster_server_add(self, _request) -> aiohttp.web.Response:
...
+        # TODO: only pass UUID
+        to_remove_ip = _request.match_info['ip']
+        to_remove_uuid = _request.match_info['uuid']
+        assert len(self.cluster.running) > 2, "Can't remove last running node"
+        initiator_ip = next(s_id for s_id in self.cluster.running if s_id != to_remove_ip)
+        logging.info("_cluster_server_remove initiator %s server %s %s", initiator_ip,
+                     to_remove_ip, to_remove_uuid)
+
+        # initate remove
+        if not await self.api.remove_node_and_wait(initiator_ip, to_remove_uuid):
```
[the remark this answers is not in the document]
Comment (alecco) [reply test/pylib/scylla_cluster.py]: Remove fails for me if the server is stopped. Is there something I'm missing?

[the remark this answers is not in the document]
Comment (alecco) [reply test/pylib/scylla_cluster.py]: Done in https://github.com/scylladb/scylladb/commit/c6d671384519aebc94439e2c64428e3e69d53441

File: test/pylib/rest_client.py (hunk 1 above)
Comment (alecco) [reply test/pylib/rest_client.py:136]: done in https://github.com/scylladb/scylladb/commit/c6d671384519aebc94439e2c64428e3e69d53441

File: test/pylib/manager_client.py [hunk 3]
```diff
@@ -86,99 +74,81 @@ async def before_test(self, test_name: str) -> None:
...
     async def server_remove(self, server_id: str) -> None:
         """Remove a specified server"""
         logger.debug("ManagerClient removing %s", server_id)
-        await self._get_text(f"/cluster/removeserver/{server_id}")
+        await self.cli.get_text(f"/cluster/removeserver/{server_id}")
         self._driver_update()
 
-    async def server_get_config(self, server_id: str) -> dict[str, object]:
-        resp = await self._get(f"/cluster/server/{server_id}/get_config")
+    async def server_get_config(self, server_id: str) -> Dict[str, object]:
```
Comment (kbr-) [inline test/pylib/manager_client.py:152]: Upgrade your mypy/python version if `dict` doesn't work for you.

File: test/pylib/rest_client.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#
+"""Asynchronous helper for Scylla REST API operations.
+"""
+import os.path
+import aiohttp
+
+
+class UnixRESTClient():
+    """An async helper for REST API operations using AF_UNIX socket"""
+    session: aiohttp.ClientSession
```
Comment (kbr-) [inline test/pylib/rest_client.py:21]: Is this field declaration necessary for anything? I see that `sock_name` is not declared, for example.

File: test/pylib/rest_client.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        return await resp.text()
+
+    async def put_json(self, resource: str, json: dict) -> aiohttp.ClientResponse:
+        """Put JSON"""
+        return await self.session.request(method="PUT", url=self._resource_uri(resource), json=json)
+
+
+class TCPRESTClient():
+    """An async helper for REST API operations"""
+    session: aiohttp.ClientSession
```
Comment (kbr-) [inline test/pylib/rest_client.py:56]: Same question as above

File: test/pylib/rest_client.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+class UnixRESTClient():
+    """An async helper for REST API operations using AF_UNIX socket"""
+    session: aiohttp.ClientSession
+
+    def __init__(self, sock_path: str):
+        self.sock_name = os.path.basename(sock_path)
+        self.session = aiohttp.ClientSession(connector = aiohttp.UnixConnector(path=sock_path))
+
+    async def stop(self) -> None:
```
Comment (kbr-) [inline test/pylib/rest_client.py:27]: This function is not used anywhere.

File: test/pylib/rest_client.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        """End session"""
+        await self.session.close()
+
+    def _resource_uri(self, resource: str) -> str:
+        # NOTE: using Python requests style URI for Unix domain sockets to avoid using "localhost"
+        return f"http+unix://{self.sock_name}{resource}"
+
+    async def get(self, resource: str) -> aiohttp.ClientResponse:
```
Comment (kbr-) [inline test/pylib/rest_client.py:35]: Please think of a way to reduce the copy pasta.
Could the bodies of `get`, `get_text` and `put_json` be refactored to reusable functions?

File: test/topology/conftest.py [hunk 8]
```diff
@@ -150,7 +150,6 @@ async def manager_internal(event_loop, request):
     port = int(request.config.getoption('port'))
     ssl = bool(request.config.getoption('ssl'))
     manager_int = ManagerClient(request.config.getoption('manager_api'), port, ssl, cluster_con)
-    await manager_int.start()
```
Comment (kbr-) [inline test/topology/conftest.py:153]: Why this change?

File: test/pylib/manager_client.py [hunk 9]
```diff
@@ -135,10 +136,17 @@ async def server_add(self) -> str:
         logger.debug("ManagerClient added %s", server_id)
         return server_id
 
-    async def server_remove(self, server_id: str) -> None:
+    async def server_remove(self, to_remove_ip: str) -> None:
```
Comment (kbr-) [inline test/pylib/manager_client.py:139]: `server_remove` and `remove_node` should be a single function.
`server_remove` on its own serves no purpose.

File: test/pylib/manager_client.py [hunk 10]
```diff
@@ -152,3 +160,11 @@ async def server_update_config(self, server_id: str, key: str, value: object) ->
                                        {"key": key, "value": value})
         if resp.status != 200:
             raise Exception(await resp.text())
+
+    async def get_server_uuid(self, server_id: str) -> None:
+        """Get server UUID through Scylla REST API"""
+        return await self.api.get_server_uuid(server_id)
+
+    async def server_wait_known(self, server_id: str, expected_server_id: str) -> bool:
```
Comment (kbr-) [inline test/pylib/manager_client.py:168]: This function is not used in this PR.
Please get rid of it and `wait_for_host_known` to make the PR easier to review.

File: test/pylib/manager_client.py [hunk 11]
```diff
@@ -135,10 +136,17 @@ async def server_add(self) -> str:
...
+    async def server_remove(self, to_remove_ip: str) -> None:
         """Remove a specified server"""
-        logger.debug("ManagerClient removing %s", server_id)
-        await self.cli.get_text(f"/cluster/removeserver/{server_id}")
+        logger.debug("ManagerClient removing server %s", to_remove_ip)
+        await self.cli.get_text(f"/cluster/remove-server/{to_remove_ip}")
+        self._driver_update()
+
+    # TODO: only pass UUID
+    async def remove_node(self, to_remove_ip: str, to_remove_uuid: str) -> None:
```
Comment (kbr-) [inline test/pylib/manager_client.py:146]: The IP given should be the IP of the server who does the remove.

File: test/pylib/rest_client.py [hunk 12]
```diff
@@ -78,3 +85,74 @@ async def put_json(self, resource: str, json: dict, host: str) \
...
+        return resp
+
+
+class ScyllaRESTAPIClient():
+    """Async Scylla REST API client"""
+
+    def __init__(self, port: int = 10000):
+        self.cli = TCPRESTClient(port)
+
+    async def stop(self):
```
Comment (kbr-) [inline test/pylib/rest_client.py:106]: You're not calling this function anywhere.

File: test/pylib/rest_client.py [hunk 13]
```diff
@@ -78,3 +85,74 @@ async def put_json(self, resource: str, json: dict, host: str) \
...
+            resp_body = await self.cli.get_text("/storage_service/removal_status",
+                                                host=initiator_ip)
+            if not server_uuid in resp_body:
+                logger.info("wait_for_remove SUCCESS %s %s", server_uuid, resp_body)
+                break
+            if time() > max_time:
+                logger.error("wait_for_remove timeout reached %s", server_uuid)
+                break
+            logger.info("wait_for_remove WAIT %s", server_uuid)
+            await asyncio.sleep(0.05)
```
Comment (kbr-) [inline test/pylib/rest_client.py:151]: That's a great way to spam the logs with `wait_for_remove WAIT ...`.
Please increase the sleep to something like `1` second.

File: test/pylib/rest_client.py [hunk 14]
```diff
@@ -78,3 +85,74 @@ async def put_json(self, resource: str, json: dict, host: str) \
...
+            logger.debug("wait_for_host_known [{%s}] not found, sleeping {sleep}", expect_server_id)
+            await asyncio.sleep(sleep)
+
+    async def remove_node(self, initiator_ip: str, server_uuid: str) -> None:
+        """Initiate remove node of server_uuid in initiator initiator_ip"""
+        resp = await self.cli.post("/storage_service/remove_node", params={"host_id": server_uuid},
+                                   host=initiator_ip)
+        logger.debug("remove_node status %s for %s", resp.status, server_uuid)
+
+    async def wait_for_remove(self, initiator_ip: str, server_uuid, timeout: int = 30) -> None:
```
Comment (kbr-) [inline test/pylib/rest_client.py:138]: Please remove the default value for `timeout`.

File: test/pylib/rest_client.py [hunk 15]
```diff
@@ -78,3 +85,74 @@ async def put_json(self, resource: str, json: dict, host: str) \
...
+            if not server_uuid in resp_body:
+                logger.info("wait_for_remove SUCCESS %s %s", server_uuid, resp_body)
+                break
+            if time() > max_time:
+                logger.error("wait_for_remove timeout reached %s", server_uuid)
+                break
+            logger.info("wait_for_remove WAIT %s", server_uuid)
+            await asyncio.sleep(0.05)
+
+    async def remove_node_and_wait(self, initiator_ip: str, server_uuid: str, timeout: int = 30) \
```
Comment (kbr-) [inline test/pylib/rest_client.py:153]: Please remove the default value for `timeout` and pass it from the call site.

File: test/pylib/scylla_cluster.py [hunk 16]
```diff
@@ -844,9 +848,33 @@ async def _cluster_server_add(self, _request) -> aiohttp.web.Response:
...
+        to_remove_ip = _request.match_info["ip"]
+        to_remove_uuid = _request.match_info["uuid"]
+        assert len(self.cluster.running) > 2, "Can't remove last running node"
+        initiator_ip = next(s_id for s_id in self.cluster.running if s_id != to_remove_ip)
+        logging.info("_cluster_remove_node initiator %s server %s %s", initiator_ip,
+                     to_remove_ip, to_remove_uuid)
+
+        # initate remove
+        try:
+            await self.api.remove_node_and_wait(initiator_ip, to_remove_uuid)
```
Comment (kbr-) [inline test/pylib/scylla_cluster.py:871]: Pass `timeout` here.
Take it from the request.
On the client side, extend the API to take the timeout and send it.
In the test send the timeout.

File: test/topology/test_topology.py [hunk 17]
```diff
@@ -45,6 +45,8 @@ async def test_remove_server_add_column(manager, random_tables):
     servers = await manager.servers()
     table = await random_tables.add_table(ncolumns=5)
     await manager.server_add()
+    server_uuid = await manager.get_server_uuid(servers[1])   # get uuid [1]
+    await manager.remove_node(servers[1], server_uuid)        # Remove   [1]
```
Comment (kbr-) [inline test/topology/test_topology.py:49]: `remove_node` should take two parameters: the IP of the removing server and the Host ID of the server being removed.

File: test/pylib/scylla_cluster.py [hunk 18]
```diff
@@ -844,9 +848,33 @@ async def _cluster_server_add(self, _request) -> aiohttp.web.Response:
...
+            return aiohttp.web.Response(status=500, text=f"{exc}")
+
+    async def _cluster_remove_node(self, _request) -> aiohttp.web.Response:
+        """Run remove node on Scylla REST API for a specified server"""
+        assert self.cluster
+        # TODO: only pass UUID
+        to_remove_ip = _request.match_info["ip"]
+        to_remove_uuid = _request.match_info["uuid"]
+        assert len(self.cluster.running) > 2, "Can't remove last running node"
+        initiator_ip = next(s_id for s_id in self.cluster.running if s_id != to_remove_ip)
```
Comment (kbr-) [inline test/pylib/scylla_cluster.py:865]: This doesn't make any sense.
The function should take two parameters: which node is removing and which node is being removed.
So `initiator_ip` should be passed as a parameter.

File: test/pylib/rest_client.py [hunk 19]
```diff
@@ -78,3 +85,74 @@ async def put_json(self, resource: str, json: dict, host: str) \
...
+    """Async Scylla REST API client"""
+
+    def __init__(self, port: int = 10000):
+        self.cli = TCPRESTClient(port)
+
+    async def stop(self):
+        """Close session"""
+        await self.cli.stop()
+
+    async def get_server_uuid(self, server_id: str):
```
Comment (kbr-) [inline test/pylib/rest_client.py:110]: Rename this to `get_server_host_id`.

File: test/pylib/manager_client.py (hunk 11 above)
Comment (kbr-) [reply test/pylib/manager_client.py:146]: also rename `to_remove_uuid` to `to_remove_host_id`.

**Review by kbr- — CHANGES_REQUESTED** [review]

**Comment by alecco** [comment]
Please ignore last push https://github.com/scylladb/scylladb/commit/4c729a221fa13f774ada5b0b90804c82388a7373

File: test/pylib/manager_client.py (hunk 3 above)
Comment (alecco) [reply test/pylib/manager_client.py]: reverted in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 5 above)
Comment (alecco) [reply test/pylib/rest_client.py]: Removed in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 6 above)
Comment (alecco) [reply test/pylib/rest_client.py]: Function now used in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/topology/conftest.py (hunk 8 above)
Comment (alecco) [reply test/topology/conftest.py:153]: Comment added in commit `HTTP client helper`
```
    While there, move aiohttp session setup back to constructors. When there
    were teardown issues it looked it could be caused by aiohttp session
    being created outside a coroutine. But this is proven not to be the case
    after recent fixes. So move it back to the ManagerClient constructor.
```

In latest version https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 4 above)
Comment (alecco) [reply test/pylib/rest_client.py]: Removed in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 7 above)
Comment (alecco) [reply test/pylib/rest_client.py]: Common base class with these methods implemented in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/manager_client.py (hunk 9 above)
Comment (alecco) [reply test/pylib/manager_client.py]: Removed old and unimplemented `server_remove` and implemented a proper one in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/manager_client.py (hunk 11 above)
Comment (alecco) [reply test/pylib/manager_client.py]: Done in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 12 above)
Comment (alecco) [reply test/pylib/rest_client.py]: `stop()` now closes `TCPRESTCli` properly in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 13 above)
Comment (alecco) [reply test/pylib/rest_client.py]: Wait methods were not needed so removed in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 15 above)
Comment (alecco) [reply test/pylib/rest_client.py]: No waits so no timeout in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 14 above)
Comment (alecco) [reply test/pylib/rest_client.py]: Wait methods were not needed so removed in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/scylla_cluster.py (hunk 16 above)
Comment (alecco) [reply test/pylib/scylla_cluster.py]: Wait methods were not needed so removed in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/topology/test_topology.py (hunk 17 above)
Comment (alecco) [reply test/topology/test_topology.py]: Manual initiator IP set in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/scylla_cluster.py (hunk 18 above)
Comment (alecco) [reply test/pylib/scylla_cluster.py]: Manual initiator IP in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

File: test/pylib/rest_client.py (hunk 19 above)
Comment (alecco) [reply test/pylib/rest_client.py]: Renamed to `get_host_id` in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/7cc3aedb8d083a86de5cf4d92fd4c3190af4413d

- All @kbr- review requests addressed
- Removed old unimplemented `server_remove`
- Track servers decommissioned and removed separately (by name) and mind returning `active_servers` vs `running_servers`
- Use `WhiteListRounRobinPolicy` with only active servers for driver connections (re-added commit from https://github.com/scylladb/scylladb/pull/11647)
- Pass `ManagerClient` to `RandomTables` and fixtures instead of directly the `cql` object so the driver can be updated when doing remove and decommission (re-added commit from https://github.com/scylladb/scylladb/pull/11647)
- Re introduce fix of `ssl` parameter clash with `ssl` module

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/861e532b02ff62bcbde5b546f411452605f23e2c

- Fixed `topology_raft_disabled/test_raft_upgrade.py` to use `ManagerClient.active_servers()`

**Comment by kbr-** [comment]
There are still conflicts

**Comment by kbr-** [comment]
> Use WhiteListRounRobinPolicy with only active servers for driver connections

> Since topology tests will take servers down, use a white list of know
running servers to create new driver connections.

It's OK for some servers to be down when the driver is reconnecting. I'm doing it all the time in upgrade tests and there's no problem.
There's no reason to use a white list of only running servers. Why did you reintroduce this change?

**Comment by kbr-** [comment]
>  test.py: topology manager handling cql connection 

The commit message for this commit doesn't make much sense to me:
> Preparing for ManagerClient to handle the driver connection, move the
use of cql fixture to manager.

ManagerClient already handles the driver connection.

> Keep topology/cql fixture to avoid clash with a top level one. To be
removed once top level conftest is gone.

There is no top level conftest for a long time now.

File: test/pylib/manager_client.py [hunk 20]
```diff
@@ -38,18 +33,11 @@ def __init__(self, sock_path: str, port: int, use_ssl: bool,
...
         self.cql: Optional[CassandraSession] = None
-        # API
-        self.sock_path = sock_path
-        self.sock_name = os.path.basename(sock_path)
-
-    async def start(self):
-        """Setup connection to Manager server"""
-        self.conn = aiohttp.UnixConnector(path=self.sock_path)
-        self.session = aiohttp.ClientSession(connector=self.conn)
+        self.cli = UnixRESTClient(sock_path)
```
Comment (kbr-) [inline test/pylib/manager_client.py:37]: Again abbreviating "client" as "cli". I already pointed this out in an earlier review. Is "client" really too long?

File: test/pylib/rest_client.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+
+    async def close(self) -> None:
+        """End session"""
+        await self.session.close()
+
+    @abstractmethod
+    def _resource_uri(self, resource: str, host: Optional[str] = None) -> str:
+        return ""
+
+    async def get(self, resource: str, host: Optional[str] = None) -> aiohttp.ClientResponse:
```
Comment (kbr-) [inline test/pylib/rest_client.py:34]: If you have an `Optional` parameter in a base class only so it can be always passed as `None` to one subclass and as not `None` to another, there's a problem with your class hierarchy.

Look: `TCPRESTClient` has this: `def _resource_uri(self, resource: str, host: Optional[str] = None)` but it's always an error to pass `host = None` for this subclass. I recommend reading about the Liskov Substitution Principle, which this hierarchy blatantly breaks.

You need to structure this differently. If you want to use inheritance the hierarchy should preserve LSP. But in this case you just want to achieve some code reuse which can be done with simple composition like extracting logic to helper functions.

File: test/topology/conftest.py (hunk 8 above)
Comment (kbr-) [reply test/topology/conftest.py:153]: I still don't understand why is it OK not to `start()` the ManagerClient at all. When is it started now?

**Review by kbr- — CHANGES_REQUESTED** [review]

**Review by kbr- — APPROVED** [review]

**Comment by alecco** [comment]
@kbr- ping
